include/cutlass/fp8_to_fp16.h

```diff
@@ -69,94 +56,81 @@ static inline ushort8 E4M3_to_FP16_chunk8(uchar8 xin) {
-  static_assert(cute::cosize_v<LayoutIn> == cute::cosize_v<LayoutOut>);
-  static_assert(cute::size_v<LayoutIn> == cute::cosize_v<LayoutIn>);
-  static_assert(cute::size_v<LayoutOut> == cute::cosize_v<LayoutOut>);
+  static_assert(cute::cosize_v<typename TensorIn::layout_type> == cute::cosize_v<typename TensorOut::layout_type>);
+  static_assert(cute::size_v<typename TensorIn::layout_type> == cute::cosize_v<typename TensorIn::layout_type>);
+  static_assert(cute::size_v<typename TensorOut::layout_type> == cute::cosize_v<typename TensorOut::layout_type>);
 
-  using SrcType = typename EngineIn::value_type;
-  using DstType = typename EngineOut::value_type;
+  using SrcType = typename TensorIn::value_type;
+  using DstType = typename TensorOut::value_type;
 
   static_assert(std::is_same_v<SrcType, uint8_t>,
                 "Expected fp8 (E4M3) input as uint8_t");
   static_assert(std::is_same_v<DstType, cute::half_t>,
                 "Expected fp16 output as half_t");
-
-  SrcType const *pSrc = reinterpret_cast<SrcType const *>(in.data());
-  DstType *pDst = reinterpret_cast<DstType *>(out.data());
-
+                
   constexpr int num_elements = decltype(size(in))::value;
 
   // TODO(Codeplay): Move conversion to NumericArrayConverter
-  if constexpr (std::is_same_v<ElementA, cute::float_e5m2_t>) {
+  if constexpr (std::is_same_v<EncodingType, cute::float_e5m2_t>) {
     static_assert(num_elements % 4 == 0,
                   "Currently, E5M2 -> FP16 conversion is only supported when "
                   "each work-item converts a multiple of 4 elements");
-    using SrcArray = cutlass::Array<uint32_t, num_elements / 4>;
-    using DstArray = cutlass::Array<uint32_t, num_elements / 2>;
-    SrcArray const *pSrcArr = reinterpret_cast<SrcArray const *>(pSrc);
-    DstArray *pDstArr = reinterpret_cast<DstArray *>(pDst);
+    auto src = cute::recast<uint32_t>(in);
+    auto dst = cute::recast<cutlass::Array<uint32_t, 2>>(out);
     // convert 4 FP8 elements at a time
-    E5M2_to_FP16<num_elements>(*pSrcArr, *pDstArr);
+    E5M2_to_FP16(src, dst);
   } else {
     // E4M3 -> FP16 conversion
     constexpr int chunk_size = 8;
     constexpr int iters = num_elements / chunk_size;
     static_assert(num_elements % chunk_size == 0,
                   "Currently, E4M3 -> FP16 conversion is only supported when "
                   "each work-item converts a multiple of 8 elements");
+    auto src = cute::recast<uchar8>(in);
+    auto dst = cute::recast<ushort8>(out);
     CUTLASS_PRAGMA_UNROLL
     for (int i = 0; i < iters; ++i) {
-      cute::intel::uchar8 src_vec;
-      CUTLASS_PRAGMA_UNROLL
-      for (int j = 0; j < chunk_size; ++j) {
-        src_vec[j] = pSrc[i * chunk_size + j];
-      }
-      cute::intel::ushort8 dst_vec;
-      dst_vec = E4M3_to_FP16_chunk8(src_vec);
-      CUTLASS_PRAGMA_UNROLL
-      for (int j = 0; j < chunk_size; ++j) {
-        reinterpret_cast<uint16_t *>(pDst)[i * chunk_size + j] = dst_vec[j];
-      }
```

Comment:
Thank you for removing all the unnecessary copies from E4M3 -> FP16 conversion! :)

Reply:
Thank you! :)